FWCore/Framework/interface/Callback.h

```diff
@@ -27,91 +27,81 @@
-               if constexpr( not std::is_same_v<produce::Null,RemainingContainerT> ) {
-                 setData<typename RemainingContainerT::head_type,
-                       typename RemainingContainerT::tail_type>(iProducts);
-               }
-            }
-         void newRecordComing() {
-            wasCalledForThisRecord_ = false;
-         }
-         
-     private:
-         Callback(const Callback&) = delete; // stop default
-         
-         const Callback& operator=(const Callback&) = delete; // stop default
-
-        std::array<void*, produce::size< TReturn >::value> proxyData_;
-         edm::propagate_const<T*> producer_;
-         method_type method_;
-         bool wasCalledForThisRecord_;
-         TDecorator decorator_;
-      };
-   }
-}
+  namespace eventsetup {
+
+    //need a virtual distructor since owner of callback only knows
+    // about the base class.  Other users of callback know the specific
+    // type
+    struct CallbackBase {
+      virtual ~CallbackBase() {}
+    };
+
+    // The default decorator that does nothing
+    template <typename TRecord> struct CallbackSimpleDecorator {
+      void pre(const TRecord&) {}
+      void post(const TRecord&) {}
+    };
+
+    template <typename T,          //producer's type
+              typename TReturn,    //return type of the producer's method
+              typename TRecord,    //the record passed in as an argument
+              typename TDecorator  //allows customization using pre/post calls
+              = CallbackSimpleDecorator<TRecord> >
+    class Callback : public CallbackBase {
+    public:
+      using method_type = TReturn (T ::*)(const TRecord&);
+
+      Callback(T* iProd, method_type iMethod, const TDecorator& iDec = TDecorator())
+          : proxyData_{}, producer_(iProd), method_(iMethod), wasCalledForThisRecord_(false), decorator_(iDec) {}
+
+      // ---------- const member functions ---------------------
+
+      // ---------- static member functions --------------------
+
+      // ---------- member functions ---------------------------
+
+      void operator()(const TRecord& iRecord) {
+        if (!wasCalledForThisRecord_) {
+          decorator_.pre(iRecord);
+          storeReturnedValues((producer_->*method_)(iRecord));
+          wasCalledForThisRecord_ = true;
+          decorator_.post(iRecord);
+        }
+      }
+
+      template <class DataT> void holdOntoPointer(DataT* iData) {
+        proxyData_[produce::find_index<TReturn, DataT>::value] = iData;
+      }
+
+      void storeReturnedValues(TReturn iReturn) {
+        //std::cout <<" storeReturnedValues "<< iReturn <<" " <<iReturn->value_ <<std::endl;
+        using type = typename produce::product_traits<TReturn>::type;
+        setData<typename type::head_type, typename type::tail_type>(iReturn);
+      }
+
+      template <class RemainingContainerT, class DataT, class ProductsT> void setData(ProductsT& iProducts) {
```

Comment:
If the options for newlines in template declarations are this or https://github.com/cms-sw/cmssw/pull/25529#discussion_r243341127, I think I'd prefer #25529's approach more (i.e. I 'd like to see a newline here more than avoiding a newline for short template declarations).